File: src/utils/signalProcessing.js
/**
 * Signal Processing Utilities for rPPG (Remote Photoplethysmography)
 * Implements CHROM algorithm for pulse extraction from facial video
 */

import {
  extractROIWithSkinFilter,
  butterworthFilter,
  chromAlgorithmImproved,
  analyzeFrequencyImproved,
  normalizeSignal,
  polynomialDetrend
} from './advancedSignalProcessing';

console.log('✅ Advanced signal processing loaded!');

// Extract RGB values from ROIs with skin filtering
export const extractROIs = (image, landmarks, canvas) => {
  const ctx = canvas.getContext('2d');
  const width = canvas.width;
  const height = canvas.height;

  // Create temporary canvas to extract pixel data
  const tempCanvas = document.createElement('canvas');
  tempCanvas.width = width;
  tempCanvas.height = height;
  const tempCtx = tempCanvas.getContext('2d');
  tempCtx.drawImage(image, 0, 0, width, height);

  const regions = {
    forehead: [10, 67, 109, 338, 297],
    leftCheek: [266, 426, 436, 416, 376],
    rightCheek: [36, 206, 216, 192, 147]
  };

  const roiData = {};

  Object.entries(regions).forEach(([region, indices]) => {
    const points = indices.map(i => ({
      x: Math.floor(landmarks[i].x * width),
      y: Math.floor(landmarks[i].y * height)
    }));

    const minX = Math.max(0, Math.min(...points.map(p => p.x)));
    const maxX = Math.min(width, Math.max(...points.map(p => p.x)));
    const minY = Math.max(0, Math.min(...points.map(p => p.y)));
    const maxY = Math.min(height, Math.max(...points.map(p => p.y)));

    const roiWidth = maxX - minX;
    const roiHeight = maxY - minY;

    if (roiWidth > 0 && roiHeight > 0) {
      const imageData = tempCtx.getImageData(minX, minY, roiWidth, roiHeight);

      // Use advanced skin filtering
      const rgb = extractROIWithSkinFilter(imageData);

      if (rgb) {
        roiData[region] = rgb;
      }
    }
  });

  return Object.keys(roiData).length === 3 ? roiData : null;
};

// Process signal using improved algorithms
const processSignal = (rgbBuffer) => {
  // Use improved CHROM algorithm
  const signal = chromAlgorithmImproved(rgbBuffer);
  if (!signal || signal.length < 60) return null;

  // Polynomial detrending
  const detrended = polynomialDetrend(signal);

  // Butterworth bandpass filter
  const filtered = butterworthFilter(detrended, 0.7, 4.0, 30);

  // Normalize
  const normalized = normalizeSignal(filtered);

  return normalized;
};

// Process rPPG signal from RGB buffer
export const processRPPGSignal = (rgbBuffer) => {
  const regions = ['forehead', 'leftCheek', 'rightCheek'];
  const signals = {};
  const freqAnalysis = {};

  // Check if we have enough data and valid skin ratios
  let validRegions = 0;
  const skinRatios = {};

  regions.forEach(region => {
    const buffer = rgbBuffer[region];
    if (buffer && buffer.length > 0) {
      const avgSkinRatio = buffer.reduce((sum, rgb) => sum + (rgb.skinRatio || 0), 0) / buffer.length;
      skinRatios[region] = avgSkinRatio;
      if (avgSkinRatio > 0.3) { // At least 30% skin pixels
        validRegions++;
      }
    }
  });

  console.log('🎨 Skin ratios:', skinRatios, 'Valid regions:', validRegions);

  if (validRegions < 2) {
    // Not enough valid skin regions
    console.log('⚠️ Insufficient skin detection');
    return {
      signals: {},
      freqAnalysis: {},
      heartRate: 0,
      consistency: 0,
      rawSignal: [],
      skinQuality: validRegions / 3
    };
  }

  // Process each ROI with improved algorithms
  regions.forEach(region => {
    const signal = processSignal(rgbBuffer[region]);
    if (signal && signal.length > 0) {
      const freq = analyzeFrequencyImproved(signal, 30);

      signals[region] = signal;
      freqAnalysis[region] = freq;

      if (region === 'forehead') {
        console.log('📊 Forehead analysis:', {
          signalLength: signal.length,
          dominantFreq: freq.dominantFreq.toFixed(3),
          dominantFreqHz: freq.dominantFreq.toFixed(3) + ' Hz',
          estimatedBPM: (freq.dominantFreq * 60).toFixed(1) + ' BPM',
          snr: freq.snr.toFixed(2),
          power: freq.power.toFixed(4),
          bandPower: freq.bandPower?.toFixed(4),
          signalMin: Math.min(...signal).toFixed(3),
          signalMax: Math.max(...signal).toFixed(3),
          signalMean: (signal.reduce((a,b) => a+b, 0) / signal.length).toFixed(3)
        });
      }
    }
  });

  // Calculate average heart rate with outlier rejection
  const heartRates = regions
    .map(r => freqAnalysis[r]?.dominantFreq || 0)
    .filter(f => f > 0)
    .map(f => f * 60); // Convert Hz to BPM

  // Remove outliers (values > 2 std dev from median)
  if (heartRates.length > 1) {
    const median = heartRates.sort((a, b) => a - b)[Math.floor(heartRates.length / 2)];
    const validHRs = heartRates.filter(hr => Math.abs(hr - median) < 40); // Within 40 BPM of median

    const avgHeartRate = validHRs.length > 0
      ? validHRs.reduce((sum, hr) => sum + hr, 0) / validHRs.length
      : 0;

    // Calculate cross-ROI consistency
    const consistency = calculateCrossROIConsistency(freqAnalysis);

    const confidenceResult = calculateConfidence({
      freqAnalysis,
      heartRate: Math.round(avgHeartRate),
      consistency,
      signals
    });

    console.log('💓 Heart rate analysis:', {
      allHRs: heartRates.map(hr => Math.round(hr)),
      validHRs: validHRs.map(hr => Math.round(hr)),
      finalHR: Math.round(avgHeartRate),
      consistency: consistency.toFixed(2)
    });

    console.log('🎯 Confidence breakdown:', {
      overall: (confidenceResult.overall * 100).toFixed(1) + '%',
      signalQuality: (confidenceResult.signalQuality * 100).toFixed(1) + '%',
      roiConsistency: (confidenceResult.roiConsistency * 100).toFixed(1) + '%',
      avgSNR: confidenceResult.snr.toFixed(2),
      hrPlausibility: (confidenceResult.heartRatePlausibility * 100).toFixed(1) + '%'
    });

    return {
      signals,
      freqAnalysis,
      heartRate: Math.round(avgHeartRate),
      consistency,
      rawSignal: signals.forehead || [],
      skinQuality: validRegions / 3
    };
  }

  return {
    signals: {},
    freqAnalysis: {},
    heartRate: 0,
    consistency: 0,
    rawSignal: [],
    skinQuality: validRegions / 3
  };
};

// Calculate cross-ROI consistency
const calculateCrossROIConsistency = (freqAnalysis) => {
  const regions = ['forehead', 'leftCheek', 'rightCheek'];
  const freqs = regions
    .map(r => freqAnalysis[r]?.dominantFreq || 0)
    .filter(f => f > 0);

  if (freqs.length < 2) return 0;

  const mean = freqs.reduce((sum, f) => sum + f, 0) / freqs.length;
  const variance = freqs.reduce((sum, f) => sum + Math.pow(f - mean, 2), 0) / freqs.length;
  const stdDev = Math.sqrt(variance);

  // Low standard deviation = high consistency
  const consistency = Math.max(0, 1 - (stdDev / mean) * 5);
  return consistency;
};

// Calculate overall confidence score
export const calculateConfidence = (signalData) => {
  const { freqAnalysis, heartRate, consistency, skinQuality = 0.5 } = signalData;

  // Average SNR across ROIs
  const snrValues = Object.values(freqAnalysis)
    .map(fa => fa.snr)
    .filter(snr => snr > 0);

  const avgSNR = snrValues.length > 0
    ? snrValues.reduce((sum, snr) => sum + snr, 0) / snrValues.length
    : 0;

  // Normalize SNR to 0-1 scale (webcam-realistic threshold)
  const snrScore = Math.min(1, avgSNR / 6); // Lowered to 6 for webcam conditions

  // Heart rate plausibility
  let hrScore = 0;
  if (heartRate >= 50 && heartRate <= 120) {
    hrScore = 1.0; // Normal resting/moderate HR
  } else if (heartRate >= 40 && heartRate < 50) {
    hrScore = 0.7; // Low but possible
  } else if (heartRate > 120 && heartRate <= 150) {
    hrScore = 0.7; // High but possible
  } else if (heartRate > 150 && heartRate <= 180) {
    hrScore = 0.3; // Extreme
  } else if (heartRate > 180 || heartRate < 40 || heartRate === 0) {
    hrScore = 0; // Impossible or no signal
  }

  // High consistency with low SNR is suspicious (photos/flat signals)
  let consistencyScore = consistency;
  if (consistency > 0.95 && avgSNR < 3) {
    consistencyScore = 0.2; // Too perfect = likely photo
  } else if (consistency > 0.85 && avgSNR < 2) {
    consistencyScore = 0.15;
  }

  // Band power check
  const bandPowerValues = Object.values(freqAnalysis)
    .map(fa => fa.bandPower || 0)
    .filter(p => p > 0);
  const avgBandPower = bandPowerValues.length > 0
    ? bandPowerValues.reduce((sum, p) => sum + p, 0) / bandPowerValues.length
    : 0;
  const bandPowerScore = Math.min(1, avgBandPower / 25); // More forgiving

  // Periodicity check
  const powerValues = Object.values(freqAnalysis)
    .map(fa => fa.power)
    .filter(p => p > 0);
  const avgPower = powerValues.length > 0
    ? powerValues.reduce((sum, p) => sum + p, 0) / powerValues.length
    : 0;
  const periodicityScore = Math.min(1, avgPower / 40); // More forgiving

  // Skin quality check
  const skinScore = Math.max(0, Math.min(1, skinQuality));

  // Signal quality - webcam-tuned weights
  const signalQuality = (
    snrScore * 0.35 +
    bandPowerScore * 0.25 +
    periodicityScore * 0.2 +
    skinScore * 0.2
  );

  // Overall confidence - emphasize HR plausibility (key discriminator)
  let overall = (
    signalQuality * 0.4 +
    hrScore * 0.3 +
    consistencyScore * 0.2 +
    periodicityScore * 0.1
  );

  // CRITICAL: Photos/spoofs must have implausible HR to be rejected
  // Only boost if HR is genuinely plausible
  if (hrScore >= 0.7 && avgSNR > 2.0) {
    overall = Math.min(1, overall * 1.6); // 60% boost only with good signal
  }

  // HEAVY penalties for clear spoofs - these must dominate
  if (heartRate > 180 || heartRate < 40 || heartRate === 0) {
    overall *= 0.1; // 90% penalty for impossible/missing HR
  }
  if (avgSNR < 2.0) {
    overall *= 0.3; // 70% penalty for very weak signal
  }
  if (avgSNR < 3.0 && consistency > 0.9) {
    // Suspicious: too consistent with weak signal (photo characteristic)
    overall *= 0.2; // 80% penalty
  }

  // Bonus for strong coherent signal (only applies after penalties)
  if (hrScore >= 0.9 && avgSNR > 4.0 && skinQuality > 0.5) {
    overall = Math.min(1, overall * 1.2); // 20% bonus for excellent signal
  }

  return {
    overall: Math.max(0, Math.min(1, overall)),
    signalQuality: Math.max(0, Math.min(1, signalQuality)),
    roiConsistency: Math.max(0, Math.min(1, consistency)),
    snr: avgSNR,
    heartRatePlausibility: hrScore,
    periodicityScore,
    bandPowerScore,
    skinQuality: skinScore
  };
};

File: android-app/app/src/main/java/com/veyrax/pulseverify/core/SignalProcessing.kt
package com.veyrax.pulseverify.core

import kotlin.math.*

/**
 * Advanced Signal Processing for rPPG
 * Implements CHROM algorithm and frequency analysis for pulse extraction
 */
object SignalProcessing {

    data class RGBValue(
        val r: Float,
        val g: Float,
        val b: Float,
        val skinRatio: Float = 0f
    )

    data class FrequencyAnalysis(
        val dominantFreq: Float = 0f,
        val snr: Float = 0f,
        val power: Float = 0f,
        val bandPower: Float = 0f,
        val peakProminence: Float = 1f
    )

    data class SignalResult(
        val signals: Map<String, List<Float>>,
        val freqAnalysis: Map<String, FrequencyAnalysis>,
        val heartRate: Int,
        val consistency: Float,
        val rawSignal: List<Float>,
        val skinQuality: Float
    )

    // ==================== SKIN DETECTION ====================

    /**
     * Detect if a pixel is skin tone using robust RGB thresholds
     */
    fun isSkinPixel(r: Int, g: Int, b: Int): Boolean {
        // Basic RGB bounds
        if (r < 40 || g < 30 || b < 15) return false
        if (r > 250 && g > 250 && b > 250) return false // Glare/overexposed

        // R should be the dominant component or close to G, R > B
        if (r < g - 10 || r < b) return false

        // Contrast between max and min channel
        val maxC = max(r, max(g, b))
        val minC = min(r, min(g, b))
        if (maxC - minC < 5) return false

        return true
    }

    /**
     * Extract ROI with skin filtering - only average skin pixels
     */
    fun extractROIWithSkinFilter(pixels: IntArray): RGBValue? {
        var r = 0f
        var g = 0f
        var b = 0f
        var skinPixelCount = 0
        val totalPixels = pixels.size

        for (pixel in pixels) {
            val red = (pixel shr 16) and 0xFF
            val green = (pixel shr 8) and 0xFF
            val blue = pixel and 0xFF

            if (isSkinPixel(red, green, blue)) {
                r += red
                g += green
                b += blue
                skinPixelCount++
            }
        }

        val skinRatio = skinPixelCount.toFloat() / totalPixels

        if (skinPixelCount < 5) {
            return null
        }

        return RGBValue(
            r = r / skinPixelCount,
            g = g / skinPixelCount,
            b = b / skinPixelCount,
            skinRatio = skinRatio
        )
    }

    // ==================== SIGNAL NORMALIZATION ====================

    /**
     * Normalize signal to zero mean and unit variance
     */
    fun normalizeSignal(signal: List<Float>): List<Float> {
        if (signal.isEmpty()) return emptyList()

        val mean = signal.average().toFloat()
        val variance = signal.map { (it - mean).pow(2) }.average().toFloat()
        val stdDev = sqrt(variance)

        if (stdDev < 1e-10f) return signal.map { 0f }

        return signal.map { (it - mean) / stdDev }
    }

    /**
     * Polynomial detrending - removes slow drifts
     */
    fun polynomialDetrend(signal: List<Float>): List<Float> {
        if (signal.size < 10) return signal

        val n = signal.size
        val x = List(n) { it.toFloat() }

        val xMean = x.average().toFloat()
        val yMean = signal.average().toFloat()

        var numerator = 0f
        var denominator = 0f

        for (i in 0 until n) {
            numerator += (x[i] - xMean) * (signal[i] - yMean)
            denominator += (x[i] - xMean).pow(2)
        }

        val slope = numerator / denominator
        val intercept = yMean - slope * xMean

        return signal.mapIndexed { i, value ->
            value - (slope * i + intercept)
        }
    }

    // ==================== CHROM ALGORITHM ====================

    /**
     * CHROM algorithm with proper color space transformation
     */
    fun chromAlgorithm(rgbBuffer: List<RGBValue>): List<Float>? {
        if (rgbBuffer.size < 30) return null

        val signal = mutableListOf<Float>()
        val validIndices = rgbBuffer.mapIndexedNotNull { index, rgb ->
            if (rgb.skinRatio > 0.3f) index else null
        }

        if (validIndices.size < 30) return null

        for (i in validIndices) {
            val rgb = rgbBuffer[i]
            val mean = (rgb.r + rgb.g + rgb.b) / 3f
            if (mean < 10f) continue

            val rNorm = rgb.r / mean
            val gNorm = rgb.g / mean
            val bNorm = rgb.b / mean

            // CHROM color space transformation
            val xs = 3f * rNorm - 2f * gNorm
            val ys = 1.5f * rNorm + gNorm - 1.5f * bNorm

            val alpha = 1.0f
            val pulse = xs - alpha * ys

            signal.add(pulse)
        }

        if (signal.size < 30) return null
        return signal
    }

    // ==================== BUTTERWORTH BANDPASS FILTER ====================

    /**
     * Butterworth-like bandpass filter
     * Frequency range: 0.7 Hz to 4 Hz (42-240 BPM)
     */
    fun butterworthFilter(
        signal: List<Float>,
        lowCut: Float = 0.7f,
        highCut: Float = 4.0f,
        fps: Float = 30f
    ): List<Float> {
        if (signal.size < 10) return signal

        // Step 1: Detrend
        var filtered = polynomialDetrend(signal)

        // Step 2: High-pass filter
        val hpWindowSize = max(3, floor(fps / lowCut / 2).toInt())
        filtered = highPassFilter(filtered, hpWindowSize)

        // Step 3: Low-pass filter
        val lpWindowSize = max(3, floor(fps / highCut / 2).toInt())
        filtered = lowPassFilter(filtered, lpWindowSize)

        // Step 4: Normalize
        filtered = normalizeSignal(filtered)

        return filtered
    }

    private fun highPassFilter(signal: List<Float>, windowSize: Int): List<Float> {
        return signal.mapIndexed { i, value ->
            val start = max(0, i - windowSize)
            val end = min(signal.size, i + windowSize + 1)
            val window = signal.subList(start, end)
            val avg = window.average().toFloat()
            value - avg
        }
    }

    private fun lowPassFilter(signal: List<Float>, windowSize: Int): List<Float> {
        return signal.mapIndexed { i, _ ->
            val start = max(0, i - windowSize)
            val end = min(signal.size, i + windowSize + 1)
            val window = signal.subList(start, end)
            window.average().toFloat()
        }
    }

    // ==================== FREQUENCY ANALYSIS ====================

    /**
     * Frequency analysis with autocorrelation for periodicity detection
     */
    fun analyzeFrequency(signal: List<Float>, fps: Float = 30f): FrequencyAnalysis {
        if (signal.size < 64) {
            return FrequencyAnalysis()
        }

        val windowSize = min(512, 2.0.pow(floor(ln(signal.size.toDouble()) / ln(2.0))).toInt())
        val windowed = signal.takeLast(windowSize)

        // Apply Hamming window
        val hammingWindow = windowed.mapIndexed { i, value ->
            val windowVal = 0.54f - 0.46f * cos((2f * PI.toFloat() * i) / (windowSize - 1))
            value * windowVal
        }

        // Compute autocorrelation
        val autocorr = computeAutocorrelation(hammingWindow, fps)

        // Find dominant frequency in heart rate range (0.7 - 4 Hz)
        val minLag = floor(fps / 4.0f).toInt()  // Max 240 BPM
        val maxLag = floor(fps / 0.7f).toInt()  // Min 42 BPM

        var maxPower = 0f
        var dominantLag = 0
        var secondPeakPower = 0f

        for (i in minLag until min(maxLag, autocorr.size)) {
            if (autocorr[i] > maxPower) {
                secondPeakPower = maxPower
                maxPower = autocorr[i]
                dominantLag = i
            } else if (autocorr[i] > secondPeakPower) {
                secondPeakPower = autocorr[i]
            }
        }

        val dominantFreq = if (dominantLag > 0) fps / dominantLag else 0f

        // Calculate noise power
        val noisePower = if (minLag > 0) {
            autocorr.take(minLag).map { abs(it) }.average().toFloat()
        } else 0f

        val snr = if (noisePower > 0) maxPower / noisePower else 0f

        // Band power
        val bandPower = autocorr.subList(minLag, min(maxLag, autocorr.size))
            .map { abs(it) }
            .sum()

        val peakProminence = if (secondPeakPower > 0) maxPower / secondPeakPower else 1f

        return FrequencyAnalysis(
            dominantFreq = dominantFreq,
            snr = snr,
            power = maxPower,
            bandPower = bandPower,
            peakProminence = peakProminence
        )
    }

    private fun computeAutocorrelation(signal: List<Float>, fps: Float): List<Float> {
        val n = signal.size
        val maxLag = min(floor(n / 2f).toInt(), floor(fps / 0.5f).toInt())
        val autocorr = mutableListOf<Float>()

        for (lag in 0 until maxLag) {
            var sum = 0f
            var count = 0

            for (i in 0 until n - lag) {
                sum += signal[i] * signal[i + lag]
                count++
            }

            autocorr.add(sum / count)
        }

        // Normalize by zero-lag value
        val maxVal = autocorr.firstOrNull() ?: 0f
        return if (maxVal > 0) {
            autocorr.map { it / maxVal }
        } else {
            autocorr
        }
    }

    // ==================== SIGNAL PROCESSING ====================

    /**
     * Process signal using CHROM + filtering
     */
    fun processSignal(rgbBuffer: List<RGBValue>): List<Float>? {
        val signal = chromAlgorithm(rgbBuffer) ?: return null
        if (signal.size < 60) return null

        val detrended = polynomialDetrend(signal)
        val filtered = butterworthFilter(detrended, 0.7f, 4.0f, 30f)
        return normalizeSignal(filtered)
    }

    /**
     * Process rPPG signal from RGB buffer
     */
    fun processRPPGSignal(rgbBuffer: Map<String, List<RGBValue>>): SignalResult {
        val regions = listOf("forehead", "leftCheek", "rightCheek")
        val signals = mutableMapOf<String, List<Float>>()
        val freqAnalysis = mutableMapOf<String, FrequencyAnalysis>()

        // Check skin quality
        var validRegions = 0
        val skinRatios = mutableMapOf<String, Float>()

        regions.forEach { region ->
            val buffer = rgbBuffer[region]
            if (buffer != null && buffer.isNotEmpty()) {
                val avgSkinRatio = buffer.map { it.skinRatio }.average().toFloat()
                skinRatios[region] = avgSkinRatio
                if (avgSkinRatio > 0.3f) {
                    validRegions++
                }
            }
        }

        if (validRegions < 2) {
            return SignalResult(
                signals = emptyMap(),
                freqAnalysis = emptyMap(),
                heartRate = 0,
                consistency = 0f,
                rawSignal = emptyList(),
                skinQuality = validRegions / 3f
            )
        }

        // Process each ROI
        regions.forEach { region ->
            val buffer = rgbBuffer[region]
            if (buffer != null) {
                val signal = processSignal(buffer)
                if (signal != null && signal.isNotEmpty()) {
                    val freq = analyzeFrequency(signal, 30f)
                    signals[region] = signal
                    freqAnalysis[region] = freq
                }
            }
        }

        // Calculate average heart rate with outlier rejection
        val heartRates = regions.mapNotNull { region ->
            freqAnalysis[region]?.dominantFreq?.let { it * 60f }
        }.filter { it > 0 }

        val avgHeartRate = if (heartRates.size > 1) {
            val sorted = heartRates.sorted()
            val median = sorted[sorted.size / 2]
            val validHRs = heartRates.filter { abs(it - median) < 40f }
            if (validHRs.isNotEmpty()) {
                validHRs.average().toFloat()
            } else 0f
        } else 0f

        // Calculate cross-ROI consistency
        val consistency = calculateCrossROIConsistency(freqAnalysis)

        return SignalResult(
            signals = signals,
            freqAnalysis = freqAnalysis,
            heartRate = avgHeartRate.roundToInt(),
            consistency = consistency,
            rawSignal = signals["forehead"] ?: emptyList(),
            skinQuality = validRegions / 3f
        )
    }

    /**
     * Calculate cross-ROI consistency
     */
    private fun calculateCrossROIConsistency(freqAnalysis: Map<String, FrequencyAnalysis>): Float {
        val freqs = freqAnalysis.values.map { it.dominantFreq }.filter { it > 0 }
        if (freqs.size < 2) return 0f

        val mean = freqs.average().toFloat()
        val variance = freqs.map { (it - mean).pow(2) }.average().toFloat()
        val stdDev = sqrt(variance)

        // Low standard deviation = high consistency
        return max(0f, 1f - (stdDev / mean) * 5f)
    }

    /**
     * Calculate overall confidence score
     */
    fun calculateConfidence(signalResult: SignalResult): ConfidenceResult {
        val freqAnalysis = signalResult.freqAnalysis
        val heartRate = signalResult.heartRate
        val consistency = signalResult.consistency
        val skinQuality = signalResult.skinQuality

        // Average SNR across ROIs
        val snrValues = freqAnalysis.values.map { it.snr }.filter { it > 0 }
        val avgSNR = if (snrValues.isNotEmpty()) {
            snrValues.average().toFloat()
        } else 0f

        val snrScore = min(1f, avgSNR / 6f)

        // Heart rate plausibility
        val hrScore = when {
            heartRate in 50..120 -> 1.0f
            heartRate in 40..49 -> 0.7f
            heartRate in 121..150 -> 0.7f
            heartRate in 151..180 -> 0.3f
            else -> 0f
        }

        // Consistency score (adjusted for photo detection)
        var consistencyScore = consistency
        if (consistency > 0.95f && avgSNR < 3f) {
            consistencyScore = 0.2f  // Too perfect = likely photo
        } else if (consistency > 0.85f && avgSNR < 2f) {
            consistencyScore = 0.15f
        }

        // Band power
        val bandPowerValues = freqAnalysis.values.map { it.bandPower }.filter { it > 0 }
        val avgBandPower = if (bandPowerValues.isNotEmpty()) {
            bandPowerValues.average().toFloat()
        } else 0f
        val bandPowerScore = min(1f, avgBandPower / 5f)

        // Periodicity
        val powerValues = freqAnalysis.values.map { it.power }.filter { it > 0 }
        val avgPower = if (powerValues.isNotEmpty()) {
            powerValues.average().toFloat()
        } else 0f
        val periodicityScore = min(1f, avgPower / 0.4f)

        val skinScore = max(0f, min(1f, skinQuality))

        // Signal quality
        val signalQuality = (
            snrScore * 0.35f +
            bandPowerScore * 0.25f +
            periodicityScore * 0.2f +
            skinScore * 0.2f
        )

        // Overall confidence
        var overall = (
            signalQuality * 0.4f +
            hrScore * 0.3f +
            consistencyScore * 0.2f +
            periodicityScore * 0.1f
        )

        // Boost for good signals
        if (hrScore >= 0.7f && avgSNR > 2.0f) {
            overall = min(1f, overall * 1.6f)
        }

        // Penalties for spoofs
        if (heartRate > 180 || heartRate < 40 || heartRate == 0) {
            overall *= 0.1f
        }
        if (avgSNR < 2.0f) {
            overall *= 0.3f
        }
        if (avgSNR < 3.0f && consistency > 0.9f) {
            overall *= 0.2f
        }

        // Bonus for excellent signals
        if (hrScore >= 0.9f && avgSNR > 4.0f && skinQuality > 0.5f) {
            overall = min(1f, overall * 1.2f)
        }

        return ConfidenceResult(
            overall = max(0f, min(1f, overall)),
            signalQuality = max(0f, min(1f, signalQuality)),
            roiConsistency = max(0f, min(1f, consistency)),
            snr = avgSNR,
            heartRatePlausibility = hrScore,
            periodicityScore = periodicityScore,
            bandPowerScore = bandPowerScore,
            skinQuality = skinScore
        )
    }

    data class ConfidenceResult(
        val overall: Float,
        val signalQuality: Float,
        val roiConsistency: Float,
        val snr: Float,
        val heartRatePlausibility: Float,
        val periodicityScore: Float,
        val bandPowerScore: Float,
        val skinQuality: Float
    )
}
